core/os/os2/process_linux.odin

```diff
@@ -71,25 +299,439 @@ _Sys_Process_Attributes :: struct {}
+	env: [^]cstring
+	if desc.env == nil {
+		// take this process's current environment
+		env = raw_data(export_cstring_environment(temp_allocator()))
+	} else {
+		cenv := make([]cstring, len(desc.env) + 1, temp_allocator())
+		for env, i in desc.env {
+			cenv[i] = temp_cstring(env) or_return
+		}
+		env = &cenv[0]
+	}
+
+	child_pipe_fds: [2]linux.Fd
+	if errno = linux.pipe2(&child_pipe_fds, {.CLOEXEC}); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	// TODO: This is the traditional textbook implementation with fork.
+	//       A more efficient implementation with vfork:
+	//
+	//       1. retrieve signal handlers
+	//       2. block all signals
+	//       3. allocate some stack space
+	//       4. vfork (waits for child exit or execve); In child:
+	//           a. set child signal handlers
+	//           b. set up any necessary pipes
+	//           c. execve
+	//       5. restore signal handlers
+	//
+	pid: linux.Pid
+	if pid, errno = linux.fork(); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	STDIN  :: linux.Fd(0)
+	STDOUT :: linux.Fd(1)
+	STDERR :: linux.Fd(2)
+
+	READ :: 0
+	WRITE :: 1
+
+	if pid == 0 {
+		// in child process now
+		write_errno_to_parent_and_abort :: proc(parent_fd: linux.Fd, errno: linux.Errno) -> ! {
+			error_byte: [1]u8 = { u8(errno) }
+			linux.write(parent_fd, error_byte[:])
+			intrinsics.trap()
+		}
+
+		stdin_fd: linux.Fd
+		stdout_fd: linux.Fd
+		stderr_fd: linux.Fd
+
+		if desc.stdin != nil {
+			stdin_fd = linux.Fd(fd(desc.stdin))
+		} else {
+			stdin_fd, errno = linux.open("/dev/null", {})
+			if errno != .NONE {
+				write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+			}
+		}
+
+		write_devnull: linux.Fd = -1
+
+		if desc.stdout != nil {
+			stdout_fd = linux.Fd(fd(desc.stdout))
+		} else {
+			write_devnull, errno = linux.open("/dev/null", {.WRONLY})
+			if errno != .NONE {
+				write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+			}
+			stdout_fd = write_devnull
+		}
+
+		if desc.stderr != nil {
+			stderr_fd = linux.Fd(fd(desc.stderr))
+		} else {
+			if write_devnull == -1 {
+				write_devnull, errno = linux.open("/dev/null", {.WRONLY})
+				if errno != .NONE {
+					write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+				}
+			}
+			stderr_fd = write_devnull
+		}
+
+		if _, errno = linux.dup2(stdin_fd, STDIN); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+		if _, errno = linux.dup2(stdout_fd, STDOUT); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+		if _, errno = linux.dup2(stderr_fd, STDERR); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+
+		success_byte: [1]u8
+		linux.write(child_pipe_fds[WRITE], success_byte[:])
+
+		if errno = linux.execveat(exe_fd, "", &cargs[0], env, {.AT_EMPTY_PATH}); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
```

Comment:
OMG i literally just realised that's what `CLOEXEC` is for! I thought there would be no way for a parent to know that the child has successfully summoned the child process, but turns out there is! They closed the pipe without writing anything to it.

Thank you for figuring this out, it took me some time to finally understand this.

Reply:
Ha, that was your idea.  I just wrote it.